### ISSUE TITLE
HttpClient on Linux not cancel request after the due time

### ISSUE THREAD

**Issue by grehovm** [dotnet/corefx#20527]
After researching [this](https://github.com/dotnet/corefx/issues/20461) problem. I came to realize that httpclient does not handle the hang of libcurl. And does not cancel hung requests at the due time. I tested 15000+ proxy servers and found 8 that hang even on fresh libcurl(7.52.1) and do not respond to cancellation. I can not get rid of hanging tasks. I do not know the analogs of Thread.Abort() for Tasks or HttpWebRequest.Abort() for HttpClient. This is also a problem, the inability to forcefully kill a task.
Test case:
```cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.Net.Http;
using System.Security.Authentication;
using System.Threading;
using System.Threading.Tasks;

namespace ProxyNetCoreTest
{
    public class Program
    {
        public static List<WebProxy> Proxies { get; } = new List<WebProxy>
        {
            new WebProxy("103.254.16.21", 8080),//This works, needed for control
            //
            new WebProxy("106.39.162.168", 80),
            new WebProxy("114.215.90.211", 1080),
            new WebProxy("115.159.90.132", 1080),
            new WebProxy("123.56.199.152", 3641),
            new WebProxy("123.56.200.75", 5261),
            new WebProxy("125.93.75.245", 1081),
            new WebProxy("125.93.75.246", 1081),
            new WebProxy("183.232.25.100", 3080),
        };

        public static void Main(string[] args)
        {
            var sw = new Stopwatch();
            sw.Start();
            foreach (var proxy in Proxies)
            {
                sw.Restart();
                Test(proxy).GetAwaiter().GetResult();
                Log($"{proxy.Ip}:{proxy.Port} -> Time: {sw.ElapsedMilliseconds}");
            }
            Log("---===   END   ===---");
            Console.ReadKey();
        }

        public static async Task Test(WebProxy proxy)
        {
            var url = "https://api.ipify.org?format=json";
            var handler = new HttpClientHandler
            {
                Proxy = proxy,
                UseCookies = false,
                ServerCertificateCustomValidationCallback = (message, xcert, chain, errors) => true,
                SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls,
                UseProxy = true
            };
            Log($"{proxy.Ip}:{proxy.Port} -> Start");
            using (var client = new HttpClient(handler))
            {
                try
                {
                    using (var tokenSource = new CancellationTokenSource(4000))
                    using (var request = new HttpRequestMessage(HttpMethod.Get, url))
                    using (var response = await client.SendAsync(request, HttpCompletionOption.ResponseContentRead, tokenSource.Token))
                    {
                        var data = await response.Content.ReadAsStringAsync();
                        Log($"{proxy.Ip}:{proxy.Port} -> OK: {data}");
                    }
                }
                catch (Exception e)
                {
                    Log($"{proxy.Ip}:{proxy.Port} -> Exception: {e.GetMessages()}");
                }
            }
        }

        public static void Log(string message)
        {
            Console.WriteLine($"{DateTime.Now:HH:mm:ss}: {message}");
        }

        public class WebProxy : IWebProxy
        {
            public string Ip { get; }
            public int Port { get; }

            private readonly Uri _uri;

            public WebProxy(string ip, int port)
            {
                Ip = ip;
                Port = port;
                _uri = new Uri("http://" + ip + ":" + port);
            }
            public Uri GetProxy(Uri destination) => _uri;

            public bool IsBypassed(Uri host) => false;

            public ICredentials Credentials { get; set; }
        }
    }

    public static class ExceptionExtensions
    {
        public static string GetMessages(this Exception self)
        {
            return self.Message + " | " + self.InnerException?.GetMessages();
        }
    }
}

```

**Comment by grehovm** [comment]
I wrote the program in C and it does not hang. Timeouts work out as expected.
```c
#include <stdio.h>
#include <curl/curl.h>
#include <time.h>

int main(void)
{
  CURL *curl;
  CURLcode res;

  curl_global_init(CURL_GLOBAL_DEFAULT);
  const char *proxies[] = {
    "http://103.254.16.21:8080",//THIS WORK
    "http://106.39.162.168:80",
    "http://114.215.90.211:1080",
    "http://115.159.90.132:1080",
    "http://123.56.199.152:3641",
    "http://123.56.200.75:5261",
    "http://125.93.75.245:1081",
    "http://125.93.75.246:1081",
    "http://183.232.25.100:3080"
  };
  curl = curl_easy_init();
  if(curl) {
    for(int i = 0; i < 8; i++) {
	time_t rawtime;
	struct tm * timeinfo;
	time ( &rawtime );
	timeinfo = localtime ( &rawtime );

	printf("\n%s%s\n", asctime (timeinfo), proxies[i]);
        curl_easy_setopt(curl, CURLOPT_URL, "https://api.ipify.org?format=json");
        curl_easy_setopt(curl, CURLOPT_PROXY, proxies[i]);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 4);
        /* Perform the request, res will get the return code */
        res = curl_easy_perform(curl);
        /* Check for errors */
        if(res != CURLE_OK)
          fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
	curl_easy_reset(curl);
	printf("\n");
    }
    /* always cleanup */
    curl_easy_cleanup(curl);
  }

  curl_global_cleanup();

  return 0;
}
```

```

Thu Jun  1 03:06:43 2017
http://103.254.16.21:8080
{"ip":"103.254.16.21"}

Thu Jun  1 03:06:46 2017
http://106.39.162.168:80
curl_easy_perform() failed: Failure when receiving data from the peer


Thu Jun  1 03:06:50 2017
http://114.215.90.211:1080
curl_easy_perform() failed: Failure when receiving data from the peer


Thu Jun  1 03:06:54 2017
http://115.159.90.132:1080
curl_easy_perform() failed: Failure when receiving data from the peer


Thu Jun  1 03:06:58 2017
http://123.56.199.152:3641
curl_easy_perform() failed: Failure when receiving data from the peer


Thu Jun  1 03:07:02 2017
http://123.56.200.75:5261
curl_easy_perform() failed: Failure when receiving data from the peer


Thu Jun  1 03:07:06 2017
http://125.93.75.245:1081
curl_easy_perform() failed: Timeout was reached


Thu Jun  1 03:07:10 2017
http://125.93.75.246:1081
curl_easy_perform() failed: Failure when receiving data from the peer
```

**Comment by stephentoub** [comment]
Thanks for the examples, @grehovm.  The C example isn't quite the same as what HttpClient does with libcurl; it doesn't use the easy interface, and it doesn't provide the timeout to libcurl.  Rather, it maintains its own processing loop that uses curl_multi_perform to process available work and then curl_multi_wait to wait for more work to be available, but with a short timeout.  Unfortunately, from looking at tracing output with your C# repro, it appears that the call to curl_multi_perform hangs; the cancellation request does arrive on time, and CurlHandler is prepared to cancel the request the moment curl_multi_perform returns, but it never does.

**Comment by stephentoub** [comment]
I opened an issue in the curl repo:
https://github.com/curl/curl/issues/1526

**Comment by stephentoub** [comment]
This is apparently a known libcurl issue:
https://github.com/curl/curl/issues/1526#issuecomment-305499863

**Comment by grehovm** [comment]
Should not we have the possibility of external control? Anything like Thread.Abort () or HttpWebRequest.Abort (). Now, after each launch, I have several hanging tasks. I see them, but I can't kill for lack of any api.

**Comment by stephentoub** [comment]
We're inside a synchronous call to an external native library.  Even thread abort wouldn't kill that.

**Comment by rmkerr** [comment]
This issue has been fixed in curl via [curl/curl#1547](https://github.com/curl/curl/pull/1547). We also have a workaround on our end via the new SocketsHttpHandler available in 2.1 preview. Closing!
